From 323fb5a1829ed22bf872c83abbf45fa63765a689 Mon Sep 17 00:00:00 2001
From: Vitaly Margolin <vmargolin@habana.ai>
Date: Thu, 08 May 2025 15:05:24 +0300
Subject: [PATCH] [SW-228332] serdes: write serdes partition sha384

Calculate SHA384 digest of full Serdes LFS partition
and write it to appropriate image descriptor

Change-Id: I516ac5c08c1edbb5cb6fc50b9d5abe8461c8d60f
Signed-off-by: Vitaly Margolin <vmargolin@habana.ai>
---

diff --git a/board/habanalabs/gaudi3/post-image-fit.sh b/board/habanalabs/gaudi3/post-image-fit.sh
index 90e3ef6..c503a94 100644
--- a/board/habanalabs/gaudi3/post-image-fit.sh
+++ b/board/habanalabs/gaudi3/post-image-fit.sh
@@ -5,6 +5,7 @@
 COMMON_BASE=${BR2_EXTERNAL}/board/habanalabs/common
 DEV_BASE=${BR2_EXTERNAL}/board/habanalabs/${HL_DEV}
 source ${COMMON_BASE}/image-utility.sh
+source ${DEV_BASE}/serdes-config.sh
 
 ETH_PHY_FW_HEX_FILE=${DEV_BASE}/eth_phy_fw.hex
 ETH_PHY_FW_HASH_FILE=${DEV_BASE}/eth_phy_fw.md5
@@ -109,18 +110,6 @@
 ITS_SED_PARAMS+="s/@MGMT_STARTADDR_LO@/${MGMT_STARTADDR_LO}/g;"
 ITS_SED_PARAMS+="s/@EVENT_STARTADDR_HI@/${ARC_EVENT_ICCM_HI}/g;"
 ITS_SED_PARAMS+="s/@EVENT_STARTADDR_LO@/${EVENT_STARTADDR_LO}/g;"
-
-SERDES_LFS_BIN_FILE_NAME="gaudi3-serdes-config.bin"
-SERDES_LFS_FILE="${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${SERDES_LFS_BIN_FILE_NAME}"
-SERDES_LFS_JSON_FILE="${BR2_EXTERNAL}/scripts/serdes/gaudi3_serdes_config_flash_memory.json"
-
-function prep_image_serdes_config {
-	# Generate the serdes LFS config bin file
-	${BR2_EXTERNAL}/scripts/serdes/serdes_generate_littlefs_flash_memory.py 	\
-							-o ${SERDES_LFS_FILE}		\
-							-j ${SERDES_LFS_JSON_FILE}
-	return
-}
 
 # serdes backup in Die1, Arc1 DCCM
 ITS_SED_PARAMS+="s/@D1_ARC1_DCCM_HI@/${D1_ARC1_DCCM_HI}/g;"
diff --git a/board/habanalabs/gaudi3_arc/post-image-spi.sh b/board/habanalabs/gaudi3_arc/post-image-spi.sh
index 1da72b7..5b2f15b 100755
--- a/board/habanalabs/gaudi3_arc/post-image-spi.sh
+++ b/board/habanalabs/gaudi3_arc/post-image-spi.sh
@@ -47,6 +47,7 @@
 IMAGE_TYPE_SPI=0
 
 ARCBOOT_BINFILE_ORIG=${ARCBOOT_BINFILE_PAD:-arcboot_rom-orig.bin}
+source ${BR2_EXTERNAL}/board/habanalabs/${HL_DEV}/serdes-config.sh
 
 function copy_all_keys {
 	#TODO: add functionality
@@ -387,12 +388,13 @@
 	local img_loadl=$5
 	local start_addr_h=$6
 	local start_addr_l=$7
+	local img_name=$8
 
 	# Prepare descriptor magic "PTR"0x$img_num
 	printf "%x%x%x%02x" "'P" "'T" "'R" $img_num | rev_bytes_stdin
 
 	# Prepare descriptor version
-	local ver=1
+	local ver=4
 	printf "%08x" $ver | rev_bytes_stdin
 
 	# Prepare image index (stable or modifiable)
@@ -420,13 +422,8 @@
 	printf "%08x" $img_loadl | rev_bytes_stdin
 
 	# Prepare checksum
-	local img_path="${STAGE_DIR}/empty.bin"
-	if [ ! -f $img_path ]; then
-		dd bs=4 count=1 if=/dev/zero of=$img_path
-	fi
-
-	local PRE_CKSUM=`${SHA_CHKSUM} ${img_path} | cut -d' ' -f1`
-	hl_chksum_add ${PRE_CKSUM}
+	local CKSUM=`sha384sum ${img_name}`
+	hl_chksum_add ${CKSUM}
 
 	# Prepare Stable and Modifiable image start address high
 	printf "%08x" $start_addr_h | rev_bytes_stdin
@@ -447,7 +444,7 @@
 	local flash_sector_size=0x10000
 	local s_off=$(($serdes_partition_sector_num * $flash_sector_size))
 
-	print_raw_data_image_desc "$img_num" "$s_off" "$flash_sector_size" 0 "$s_off" 0 "$s_off"
+	print_raw_data_image_desc "$img_num" "$s_off" "$flash_sector_size" 0 "$s_off" 0 "$s_off" "$SERDES_LFS_FILE"
 }
 
 function print_image_desc {
@@ -797,4 +794,4 @@
 
 	# pad arcboot rom binary to size of ROM memory
 	dd status=none conv=sync bs=128K if=${BINS_DIR}/${ARCBOOT_BINFILE_ORIG} > ${BINS_PROD_DIR}/${ARCBOOT_BINFILE}
-}
+}
\ No newline at end of file
diff --git a/board/habanalabs/gaudi3_arc/serdes-config.sh b/board/habanalabs/gaudi3_arc/serdes-config.sh
new file mode 100644
index 0000000..cf8bc31
--- /dev/null
+++ b/board/habanalabs/gaudi3_arc/serdes-config.sh
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+SERDES_LFS_BIN_FILE_NAME="gaudi3-serdes-config.bin"
+SERDES_LFS_FILE="${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${SERDES_LFS_BIN_FILE_NAME}"
+SERDES_LFS_JSON_FILE="${BR2_EXTERNAL}/scripts/serdes/gaudi3_serdes_config_flash_memory.json"
+
+function prep_image_serdes_config {
+	# Generate the serdes LFS config bin file
+	${BR2_EXTERNAL}/scripts/serdes/serdes_generate_littlefs_flash_memory.py 	\
+							-o ${SERDES_LFS_FILE}		\
+							-j ${SERDES_LFS_JSON_FILE}
+	return
+}
\ No newline at end of file
